### PR TITLE
Support for include files

### PR DESCRIPTION
**Pull request by mattiasljungstrom**
This adds @include file support. A few notes: 

- I have not added an include_dirs to the input parameters, instead the directory of the base file is used if file is not found in cwd.
- Spaces in filenames are not supported.
- Updated error messages to show correct filename/line nr. I haven't tested all of them, but the change is very similar at each error.
- Checks for include cycles.

I haven't tested this on macOS or Linux. :)

### CODE REVIEW

**Comment by mattiasljungstrom** [comment]
will fix the macOS / linux compiler errors!

**Comment by mattiasljungstrom** [comment]
The fips compilation system (SokolShader.py) will not catch changes in included shader files. Not sure best way to handle that.

**Comment by floooh** [comment]
Haven't had time to look at the PR yet, sorry...

For the fips stuff, I think the best solution is to add the include files to the sokol_shader() macro, something like:

```cmake
sokol_shader(bla.glsl glsl330 INCLUDES blub.h blob.h)
```
The dirty check in the python script can already handle multiple input files, it just doesn't know how to get those automatically.

File: src/shdc/shdc.h [hunk 1]
```diff
@@ -228,6 +238,25 @@ struct input_t {
     input_t() { };
     static input_t load_and_parse(const std::string& path);
     void dump_debug(errmsg_t::msg_format_t err_fmt) const;
+
+    errmsg_t error(int index, const std::string& msg) const {
+        if (index < lines.size()) {
```
Comment (floooh) [inline src/shdc/shdc.h]: I'm getting a warning here about comparison of integers of different signs.

File: src/shdc/shdc.h [hunk 2]
```diff
@@ -228,6 +238,25 @@ struct input_t {
...
+            const line_t& line = lines[index];
+            return errmsg_t::error(filenames[line.filename], line.index, msg);
+        }
+        else {
+            return errmsg_t::error(base_path, 0, msg);
+        }
+    };
+    errmsg_t warning(int index, const std::string& msg) const {
+        if (index < lines.size()) {
```
Comment (floooh) [inline src/shdc/shdc.h]: ...and here also a warning about comparison of signed / unsigned integrs.

File: src/shdc/input.cc [hunk 3]
```diff
@@ -398,36 +400,105 @@ static bool parse(input_t& inp) {
...
+            else {
+                // otherwise process file as normal
+                inp.lines.push_back({line, filename_index, line_index});
+            }
+        }
+        // empty lines are not added
+
+        line_index++;
     }
+}
```
Comment (floooh) [inline src/shdc/input.cc]: ...and here a warning "Control may reach end of non-void function".

**Comment by floooh** [comment]
I'm getting three warnings when compiling in Xcode11 Beta, see above.

Hmm, and when testing in sokol-samples (without changes in sokol-samples), I'm getting incomplete generated header files, all the actual content seems to be missing :)

```cpp
#pragma once
/*
    #version:2# (machine generated, don't edit!)

    Generated by sokol-shdc (https://github.com/floooh/sokol-tools)

    Overview:


    Shader descriptor structs:


    Image bind slots, use as index in sg_bindings.vs_images[] or .fs_images[]


*/
#include <stdint.h>
#include <stdbool.h>
#if !defined(SOKOL_SHDC_ALIGN)
#if defined(_MSC_VER)
#define SOKOL_SHDC_ALIGN(a) __declspec(align(a))
#else
#define SOKOL_SHDC_ALIGN(a) __attribute__((aligned(a)))
#endif
#endif
#if !defined(SOKOL_SHDC_DECL)
#if !defined(SOKOL_GFX_INCLUDED)
#error "Please include sokol_gfx.h before sdf-sapp.glsl.h"
#endif
#endif /* SOKOL_SHDC_DECL */
```

**Comment by floooh** [comment]
...hmm so far I cannot reproduce the problem when running sokol-shdc directly in the debugger (not via fips)...

**Comment by floooh** [comment]
Ah ok, seems the problem is caused by the missing return value, adding a "return true" at the end seem to fix the problem, however, if this is the case it looks like the tools doesn't properly return with an error if that method returns with false... I'll test that.

**Comment by floooh** [comment]
Yep, that's the problem. When the input_t::load_and_preprocess() function doesn't set any errors, but still doesn't return "true" (because currently it's missing), then the input_t::parse() function isn't called, and everything remains empty even though no further errors happen.

Just returning true in load_and_preprocess() fixes the problem (it's a bit "smelly" IMHO because there are two separate "success-states", one being the out_error member, and the other the return value.

File: src/shdc/input.cc (hunk 3 above)
Comment (mattiasljungstrom) [reply src/shdc/input.cc]: doh, I really don't understand why VS doesn't warn for this.

File: src/shdc/shdc.h (hunk 2 above)
Comment (mattiasljungstrom) [reply src/shdc/shdc.h]: didn't see these warnings in VS, but I'll fix them!

**Comment by mattiasljungstrom** [comment]
good, point, maybe only the error should be checked.

**Comment by mattiasljungstrom** [comment]
the same pattern is used in all the validate_xxx methods.

**Comment by floooh** [comment]
Ok, I did some local testing and it looks good here :) Tracking and displaying errors in include files works nicely.

Do you want to do more changes, or is it ready for merging?

**Comment by mattiasljungstrom** [comment]
No more changes planned, hopefully good to go! Thanks!

**Comment by floooh** [comment]
Alright, merged, going to update the exes in sokol-tools-bin now :)

**Comment by floooh** [comment]
Ok, exes have been updated.

**Pull request merged**